lib/rust/flexer/src/generate.rs

```diff
@@ -0,0 +1,531 @@
+            reader.advance_char(&mut self.bookmarks);
+            while self.run_current_state(&mut reader) == StageStatus::ExitSuccess {}
+            match self.status {
+                StageStatus::ExitFinished => LexingResult::success(
+                    mem::replace(&mut self.output,vec![])
+                ),
+                StageStatus::ExitFail => LexingResult::failure(
+                    mem::replace(&mut self.output,vec![])
+                ),
+                _ => LexingResult::partial(mem::replace(&mut self.output,vec![]))
+            }
+        }
+    };
+    Ok(tree)
+}
+
+/// Generate the function responsible for executing the lexer in its current state.
+pub fn run_current_state_function() -> ImplItem {
+    let tree:ImplItem = parse_quote! {
+        fn run_current_state<R:LazyReader>(&mut self, reader:&mut R) -> StageStatus {
+            self.status = StageStatus::Initial;
+
+            // Runs until reaching a state that no longer says to continue.
+            while let Some(next_state) = self.status.continue_as() {
+                self.logger.info(||format!("Current character is {:?}",reader.character()));
+                self.logger.info(||format!("Continuing in {:?}",next_state));
+                self.status = self.step(next_state,reader);
+
+                if reader.finished() {
+                    self.status = StageStatus::ExitFinished
+                }
+
+                if self.status.should_continue() {
+                    match reader.character().char {
+                        Ok(char) => {
+                            reader.append_result(char);
+                            self.logger.info(||format!("Result is {:?}",reader.result()));
+                        },
+                        Err(flexer::prelude::reader::Error::EndOfGroup) => {
+                            let current_state = self.current_state();
+                            let group_name    = self.groups().group(current_state).name.as_str();
+                            let err           = format!("Missing rules for state {}", group_name);
+                            self.logger.error(err.as_str());
+                            panic!(err)
+                        }
+                        Err(_) => {
+                            self.logger.error("Unexpected error.");
+                            panic!("Unexpected error!")
+                        }
+                    }
+                    reader.advance_char(&mut self.bookmarks);
+                }
+            }
+
+            self.status
+        }
+    };
+    tree
+}
+
+/// Generate the `step` function for the lexer.
+///
+/// This function is responsible for dispatching based on the current state, consuming a character,
+/// and returning the state to transition to.
+pub fn step(groups:&group::Registry) -> ImplItem {
+    let arms            = groups.all().iter().map(|g| step_match_arm(g.id.into())).collect_vec();
+    let match_expr:Expr = parse_quote! {
```

Comment:
Do we need this as a separate variable at all?
I thinkk it might be more readable if this was just inlined in the step_tree definition.
Also, likely then `step_tree` wouldn't need to be a variable and the type will be inferred from the function's return value.